File: src/id0061.py
# Licensed under the MIT License.

# Cyclic Figurate Numbers

from math import sqrt
from time import time

def math_is_cyclic(a: int, b: int) -> bool: 
    return a != b and a // 100 == b % 100

def math_is_polygonal(s: int, x: int) -> bool:
    sM2 = s - 2
    sM4 = sM2 - 2
    term = (sqrt(8 * sM2 * x + sM4 * sM4) + sM4) / (2 * sM2)

    return term == int(term)

def math_is_octagonal_or_smaller_polygonal(x: int) -> bool:
    for s in range(3, 6):
        if math_is_polygonal(s, x): return True
    
    return math_is_polygonal(7, x) or math_is_polygonal(8, x)

def math_is_permuted_polygonal(values: list) -> bool:
    flags = {}
    
    for i in range(3, 9):
        flags[i] = False
        
    for x in values:
        for s in range(8, 2, -1):
            if not math_is_polygonal(s, x): continue
            
            flags[s] = True
            break
            
    return all(flags.values())
    
def math_cyclic_polygonal_sum(polygonals: list[int]) -> int:
    for a in polygonals:
        for b in polygonals:
            if not math_is_cyclic(a, b): continue
            
            for c in polygonals:
                if not math_is_cyclic(b, c): continue
                
                for d in polygonals: 
                    if not math_is_cyclic(c, d): continue
                    
                    for e in polygonals: 
                        if not math_is_cyclic(d, e): continue
                        
                        for f in polygonals:
                            if not math_is_cyclic(e, f): continue
                            if not math_is_cyclic(f, a): continue
                            
                            values = [a, b, c, d, e, f]
                            
                            if not math_is_permuted_polygonal(values): continue
                            
                            return sum(values)

    return -1

start = time()
polygonals = []

for x in range(1000, 10000):
    if math_is_octagonal_or_smaller_polygonal(x):
        polygonals.append(x)
    
result = math_cyclic_polygonal_sum(polygonals)

print(f"0061{result:>64}    {time() - start:.6f}")

File: src/id0061.cs
// Licensed under the MIT License.

// Cyclic Figurate Numbers

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

bool IsCyclic(int a, int b)
{
    return a / 100 == b % 100;
}

bool IsPolygonal(int s, int x)
{
    double sM2 = s - 2;
    double sM4 = sM2 - 2;
    double term = (Math.Sqrt(8 * sM2 * x + sM4 * sM4) + sM4) / (2 * sM2);

    return term == (int)term;
}

bool IsOctagonalOrSmallerPolygonal(int x)
{
    for (int s = 3; s < 6; s++)
    {
        if (IsPolygonal(s, x))
        {
            return true;
        }
    }

    return IsPolygonal(7, x) || IsPolygonal(8, x);
}

bool IsPermutedPolygonal(IReadOnlyCollection<int> values)
{
    if (values.Count != 6)
    {
        return false;
    }

    int flags = 0;

    foreach (int x in values)
    {
        for (int s = 8; s >= 3; s--)
        {
            if (IsPolygonal(s, x))
            {
                flags |= (1 << s);

                break;
            }
        }
    }

    return flags == 0b111111000;
}

int CyclicPolygonalSum()
{
    HashSet<int> polygonals = new HashSet<int>();

    for (int x = 1000; x < 10000; x++)
    {
        if (IsOctagonalOrSmallerPolygonal(x))
        {
            polygonals.Add(x);
        }
    }

    foreach (int a in polygonals)
    {
        foreach (int b in polygonals)
        {
            if (!IsCyclic(a, b))
            {
                continue;
            }

            foreach (int c in polygonals)
            {
                if (!IsCyclic(b, c))
                {
                    continue;
                }

                foreach (int d in polygonals)
                {
                    if (!IsCyclic(c, d))
                    {
                        continue;
                    }

                    foreach (int e in polygonals)
                    {
                        if (!IsCyclic(d, e))
                        {
                            continue;
                        }

                        foreach (int f in polygonals)
                        {
                            if (!IsCyclic(e, f) || !IsCyclic(f, a))
                            {
                                continue;
                            }

                            HashSet<int> values = new HashSet<int>()
                            {
                                a, b, c, d, e, f
                            };

                            if (!IsPermutedPolygonal(values))
                            {
                                continue;
                            }

                            return values.Sum();
                        }
                    }
                }
            }
        }
    }

    return -1;
}

Stopwatch start = Stopwatch.StartNew();

int result = CyclicPolygonalSum();

start.Stop();

Console.WriteLine("0061{0,64}    {1:f6}", result, start.Elapsed.TotalSeconds);
